pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java

```diff
@@ -610,6 +611,40 @@ private static Expression toExpression(SqlNode node) {
         }
         asFuncExpr.getFunctionCall().addToOperands(RequestUtils.getIdentifierExpression(aliasName));
         return asFuncExpr;
+      case CASE:
+        // CASE WHEN Statement is model as a function with variable length parameters.
+        // Assume N is number of WHEN Statements, total number of parameters is (2 * N + 1).
+        // - N: Convert each WHEN Statement into a function Expression;
+        // - N: Convert each THEN Statement into an Expression;
+        // - 1: Convert ELSE Statement into an Expression.
+        SqlCase caseSqlNode = (SqlCase) node;
+        SqlNodeList whenOperands = caseSqlNode.getWhenOperands();
+        SqlNodeList thenOperands = caseSqlNode.getThenOperands();
+        SqlNode elseOperand = caseSqlNode.getElseOperand();
+        Expression caseFuncExpr = RequestUtils.getFunctionExpression(SqlKind.CASE.name());
+        for (SqlNode whenSqlNode : whenOperands.getList()) {
+          Expression whenExpression = toExpression(whenSqlNode);
+          if (isAggregateExpression(whenExpression)) {
+            throw new SqlCompilationException(
+                "Aggregation functions inside WHEN Clause is not supported - " + whenSqlNode);
+          }
+          caseFuncExpr.getFunctionCall().addToOperands(whenExpression);
+        }
+        for (SqlNode thenSqlNode : thenOperands.getList()) {
```

Comment:
- Can THEN part itself have a CASE statement or we can assume they will always be literal expressions?

- Right now we are only supporting as part of SELECT list?

Reply:
1. For THEN part, we support both transform function and literal, technically it could be another CASE statement inside.
2. Since it's modeled as a transform function, then it could be used any where allowing this. e.g. predicate, group bys.